common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/PKeyAuthChallenge.java

```diff
@@ -87,6 +96,78 @@ public PKeyAuthChallenge(@NonNull final String redirectUri) throws ClientExcepti
         mSubmitUrl = parameters.get(PKeyAuthChallengeHandler.RequestField.SubmitUrl.name());
     }
 
+    public PKeyAuthChallenge(@NonNull final String header, @NonNull final String uri)
+            throws ClientException, UnsupportedEncodingException {
+        validateHeaderForPkeyAuthChallenge(header);
```

Comment:
Can we modularize this into helper methods rather than having this whole code block in a constructor ?

Reply:
Addressed.

Reply:
Hmm.... looks like there's still a lot required to construct this object...

@heidijinxujia How would you feel about creating a `Factory` class to create these challenges? I am thinking something like this...

```java
PkeyAuthChallengeFactoryfactory = new PkeyAuthChallengeFactory();
PkeyAuthChallenge challenge = factory.newChallenge(header, uri);
```

Inside the implementation of `newChallenge()` we would do all of the parsing of `submitUrl`, `nonce`, `version`, `context`, `authorities`, and `thumbprint`. Then, the factory would call a simple constructor on the challenge...
```java
class PkeyAuthChallengeFactory {
    PkeyAuthChallenge newChallenge(@NonNull final String header, @NonNull final String uri) {
        // Call utility methods of the factory....
        return new PkeyAuthChallenge(submitUrl, nonce, version, context, thumbprint, authorities);
    }

    // Utility functions to support creating the challenge

    private static getNonce(@NonNull Map<String, String> headerItems) {
        return headerItems.get(PKeyAuthChallengeHandler.RequestField.Nonce.name().toLowerCase(Locale.US));
    }
}
```

Reply:
@iambmelt Great idea, addressed & refactored.